lib/rpmsg/rpmsg_virtio.c

```diff
@@ -492,6 +492,11 @@ static int rpmsg_virtio_ns_callback(struct rpmsg_endpoint *ept, void *data,
 			_ept->dest_addr = dest;
 			metal_mutex_release(&rdev->lock);
 		}
+	} else { /* RPMSG_NS_CREATE_ACK */
+		/* save the received destination address */
+		if (_ept)
+			_ept->dest_addr = dest;
+		metal_mutex_release(&rdev->lock);
```

Comment:
Here you can not ensure when you will receive the ack. And there is no callback to inform about the ack reception.
that's means that your application will have to poll on ept->dest_addr. So we fall back in the issue that the remote application have to be ready...
So for me something is still missing here. The ack solution needs a trigger to inform application that everything is ready for the communication. But this will also not ensure that the remote application is ready but only the rpmsg remote handler itself.
So a first message from the server seems to me the best solution ( even if not the cleanest) to ensure that everyone is ready for the communication.

Now an alternative is to invert the initialization and use ns_bind_cb .
For instance if processor A want to send logs to processor B through an RPMsg service. A solution is that the processor B create the endpoint and so sent the NS announcement to the processor A that creates associated endpoint and start the communication.

But i suppose that you issue is the communication with Linux. In this case the issue seems more on Linux side, as you want to generate a NS service from a Linux rpmsg driver (that is probed by the rpmsg _bus). I never tried to generate a NS announcement from the Linux side, but seems that by fixing the source address you can generate a ns announcement.

Anyway this subject is interesting, and should be discussed in community to identify if some other companies have same need and to find a solution common between Linux and OpenAMP. I will mention it in next openamp-rp meeting.

Reply:
> Here you can not ensure when you will receive the ack. And there is no callback to inform about the ack reception.
> that's means that your application will have to poll on ept->dest_addr. So we fall back in the issue that the remote application have to be ready...

We add callback before, but find nobody(outside OpenAMP expert) monitor the notification. Most people call rpmsg_send immediately after rpms_create_ept return(just like socket/sendto/recvfrom), and find that it work most time but not always(depends on how fast the peer response). Since it is a very tough bug to reproduce, debug and fix, we decide to switch to poll approach like this patch:
https://github.com/OpenAMP/open-amp/pull/204/commits/fdda485c9569ba3b869ebc8e2a4f778b58972783

> So for me something is still missing here. The ack solution needs a trigger to inform application that everything is ready for the communication.

The callback can be added late for the advanced usecase if needed. But for the most case, the callback just make the simple thing too complicate to use correctly.

> But this will also not ensure that the remote application is ready but only the rpmsg remote handler itself.

No, ACK is sent only after the driver probe return OK. It's a common sense that the driver is fully functional after this point.

> So a first message from the server seems to me the best solution ( even if not the cleanest) to ensure that everyone is ready for the communication.

It is very unnatural that the first message must send from the server. Many server just do the action upon receiving the request(e.g. log redirect, file access, i2c/spi proxy...), how do you design the first meaningful message sent from server? I saw many rpmsg protocol define something like RPMG_CMD_DUMMY, It's a bad design that we push the internal design flaw to the application developer.

> 
> Now an alternative is to invert the initialization and use ns_bind_cb .
> For instance if processor A want to send logs to processor B through an RPMsg service. A solution is that the processor B create the endpoint and so sent the NS announcement to the processor A that creates associated endpoint and start the communication.

The remote side may want to disable some rpmsg service temporarily(.e.g. switch the log from rpmsg to serial to debug the hang issue). This alternative just make the whole process much complex(both side need sync with each other). It become more worse if we consider many RTOS developer isn't familiar with Linux kernel.

We should fix the design flaw in protocol level, it isn't a good idea to spread the workaround/alternative across the application protocol.

> 
> But i suppose that you issue is the communication with Linux. In this case the issue seems more on Linux side, as you want to generate a NS service from a Linux rpmsg driver (that is probed by the rpmsg _bus). I never tried to generate a NS announcement from the Linux side, but seems that by fixing the source address you can generate a ns announcement.
> 

At least, the current kernel code base can't proactively create th new rpmsg_device.

> Anyway this subject is interesting, and should be discussed in community to identify if some other companies have same need and to find a solution common between Linux and OpenAMP. I will mention it in next openamp-rp meeting.

Here is the patch for kernel side:
https://patchwork.kernel.org/cover/10790823/

Reply:
Same comment that for the dynamic RPMsg size, this impact the rpmsg protocol.
Let's other peoples give their opinions.
So Please sent or resent the corresponding patch on Linux side to trig discussion.
Seems that the link to the Linux patch you provide is not good. could you update it?

Reply:
Sure, I will update two kernel patch series in the next couple day.